File: playwright-typescript/pages/checkout-page.ts
import { Locator, Page } from '@playwright/test';

export class CheckoutPage {
  readonly page: Page;
  readonly iframe;
  readonly emailField: Locator;
  readonly nameField: Locator;
  readonly cardNumberField: Locator;
  readonly cardExpiryField: Locator;
  readonly cardCvcField: Locator;
  readonly regionField: Locator;
  readonly submitButton: Locator;
  readonly paymentDeclinedField: Locator;
  readonly paymentAuthenticationFailureField: Locator;

  constructor(page: Page) {
    this.page = page;
    this.iframe = page.frameLocator('#checkout-demo');
    this.emailField = this.iframe.locator('#email');
    this.nameField = this.iframe.locator('#billingName');
    this.cardNumberField = this.iframe.locator('#cardNumber');
    this.cardExpiryField = this.iframe.locator('#cardExpiry');
    this.cardCvcField = this.iframe.locator('#cardCvc');
    this.regionField = this.iframe.locator('#billingPostalCode');
    this.submitButton = this.iframe.locator('data-testid=hosted-payment-submit-button');
    // Looks like UI can get 2 texts randomly in the `decline` flow:
    // 1. 'Your card was declined. Please try a different card.'
    // 2. 'Your credit card was declined. Try paying with a debit card instead.'
    this.paymentDeclinedField = this.iframe.locator("//span[@role='alert' and contains(text(),'card was declined.')]");
    this.paymentAuthenticationFailureField = this.iframe.locator("//*[contains(text(),'We are unable to authenticate your payment method. Please choose a different payment method and try again.')]");
  }

  async goto() {
    await this.page.goto('https://checkout.stripe.dev/preview');
  }

  async enterEmail(email: string) {
    await this.emailField.fill(email);
  }

  async enterName(name: string) {
    await this.nameField.fill(name);
  }

  async enterPaymentCardInformation(number: string, expiry: string, cvc: string) {
    await this.cardNumberField.fill(number);
    await this.cardExpiryField.fill(expiry);
    await this.cardCvcField.fill(cvc);
  }

  async enterCountry(zip: string) {
    // use USA as country by default
    await this.regionField.fill(zip);
  }

  async submitPayment() {
    await this.submitButton.click();
  }
}

File: selenium-ruby/features/pages/checkout_page.rb
require_relative 'base_page'

class CheckoutPage < BasePage
  def id
    'checkout-demo'
  end

  def enter_email(email = PAYMENT_CARD_HOLDER_EMAIL[:valid])
    wait_for { email_field }.send_keys(email)
  end

  def enter_name(name = PAYMENT_CARD_HOLDER_NAME[:valid])
    wait_for { name_field }.send_keys(name)
  end

  def enter_payment_card_information(card_number = PAYMENT_CARD_NUMBER[:valid],
                                     card_expiry = PAYMENT_CARD_EXPIRATION_DATE[:valid],
                                     card_cvc = PAYMENT_CARD_CVC[:valid])
    wait_for { card_number_field }.send_keys(card_number)
    wait_for { card_expiry_field }.send_keys(card_expiry)
    wait_for { card_cvc_field }.send_keys(card_cvc)
  end

  def enter_country(zip = PAYMENT_CARD_ZIP[:valid])
    # use USA as country by default
    wait_for { region_field }.send_keys(zip)
  end

  def submit_payment
    wait_for { submit_button }.click
  end

  def payment_declined_msg
    payment_declined_field.text
  end

  def payment_error_msg
    payment_error_field.text
  end

  def collapse_test_inputs
    wait_for { test_inputs_header }.click
  end

  private

  def email_field
    find_ele('email')
  end

  def name_field
    find_ele('billingName')
  end

  def card_number_field
    find_ele('cardNumber')
  end

  def card_expiry_field
    find_ele('cardExpiry')
  end

  def card_cvc_field
    find_ele('cardCvc')
  end

  def region_field
    find_ele('billingPostalCode')
  end

  def submit_button
    find_by_class('SubmitButton-Shimmer')
  end

  def payment_declined_field
    find_by_xpath("//span[@role='alert']")
  end

  def payment_error_field
    find_by_class('ConfirmPaymentButton-Error')
  end

  def test_inputs_header
    find_by_class('TestInputs-Header')
  end
end
